snuba_sdk/request.py

```diff
@@ -0,0 +1,99 @@
+from __future__ import annotations
+
+import json
+import re
+from dataclasses import dataclass, fields
+
+from snuba_sdk.query import Query
+
+
+class InvalidRequestError(Exception):
+    pass
+
+
+class InvalidFlagError(Exception):
+    pass
+
+
+@dataclass(frozen=True)
+class Flags:
+    totals: bool | None = None
+    consistent: bool | None = None
+    turbo: bool | None = None
+    debug: bool | None = None
+    dry_run: bool | None = None
+    legacy: bool | None = None
+
+    def validate(self) -> None:
+        flags_fields = fields(self)
+        for ff in flags_fields:
+            flag = getattr(self, ff.name)
+            if flag is not None and not isinstance(flag, bool):
+                raise InvalidFlagError(f"{ff.name} must be a boolean")
+
+    def to_dict(self) -> dict[str, bool]:
+        self.validate()
+        values = {
+            "totals": self.totals,
+            "consistent": self.consistent,
+            "turbo": self.turbo,
+            "debug": self.debug,
+            "dry_run": self.dry_run,
+            "legacy": self.legacy,
+        }
+        return {f: v for f, v in values.items() if v is not None}
+
+
+FLAG_RE = re.compile(r"^[a-zA-Z0-9_\.\+\*\/:\-\[\]]*$")
+
+
+@dataclass
+class Request:
+    dataset: str
+    app_id: str
+    query: Query
+    flags: Flags | None = None
+
+    def set_query(self, query: Query) -> Request:
+        self.query = query
+        return self
+
+    def set_flags(self, flags: Flags) -> Request:
+        self.flags = flags
+        return self
```

Comment:
They allow chaining of calls (`Request.set_query(...).set_flags(...)`) and match the style of how the Query is built.

Reply:
Ok but this is not a query. This is a request. The query is trying to mimic an ORM like sqlalchemy where you do

```python3
Query(SomeEntity).where(SomeEntity.blah == "blah").limit(10)
```

That's not what a request object is. The [python requests library](https://docs.python-requests.org/en/latest/) doesn't have this pattern because it doesn't make sense for this usecase. Our code is polluted with java-like getters and setters when it doesn't need to be. It's just unnecessary code and I'd like us to stop adding it.

Reply:
I guess we can agree to disagree. I quite like the chaining pattern. I think it makes sense that if we have that pattern on one object we should support it on others as well.

Also the Query doesn't use this pattern to look like an ORM, it does it because the Query object is immutable. The setters allow for much cleaner code.

```
query = Query(...).set_select(...).set_where(...)

# vs.

query = replace(replace(Query(...), select=...), where=...)
```

Reply:
> Why is the query immutable? What does it give us?

It allows us to give some guarantee that queries cannot be made invalid after they have been validated.

Regardless, method chaining is a valid software pattern whether you agree with it or not. I can also quote the Zen of Python to argue my side:

`Flat is better than nested.`

```
query = Query(...).set_select(...).set_where(...) # flat

# vs.

query = replace(replace(Query(...), select=...), where=...) # nested
```

`There should be one-- and preferably only one --obvious way to do it.`

There is only one `set_*` method for each field, and trying to set the field directly won't work with the `Query` class. So there's only one obvious way to set fields on the `Query` class. Why would the `Request` class behave differently? That would be more than one way.

I could keep going but all of this is subjective opinion, and this PR has been blocked for 10 days because of disagreements over code style. We can debate that directly if you like (not on Github), but unless there are non-stylistic changes to this PR I'm going to merge it.

Reply:
> Why would the `Request` class behave differently?

IMO, a `Query` and a `Request` are inherently two different concepts. The former is an ordered set of 'commands', and the latter is an unordered set of 'flags'. I don't see why they should be consistent with each other in terms of how they are built.

I could see a use case for a builder pattern for a Request where we do something like:
```python3
request = RequestBuilder().set_query(query).has_totals().is_legacy().build()
```

However, our `Flags` class already bundles everything into a dataclass, what difference would it really make if we do `Request(query, flags)` vs `Request().set_query(query).set_flags(flags)`? Any new flags introduced will not change this format or grow this chain in any meaningful way, so the use case of the builder pattern here feels weak to me.

The only other reason I can see these setters making sense is if there's a ton of processing between when the query and flags are set in the request, and we'd like to avoid passing around the query or flags object around.

I don't see myself building a Request by chaining these 2 methods together (if that's the only use case for these setters) if I have the option to just directly pass these 2 arguments in the constructor.

Reply:
The builder pattern is an interesting one, but you make a fair point that the `Request` probably won't get more complex. OK I'll change this.